server/auth/registrar.go

```diff
@@ -98,6 +100,127 @@ func (auth *AuthManager) handleRegister(conn comms.Link, msg *msgjson.Message) *
 	return nil
 }
 
+// handleReinstate handles requests to the 'reinstate' route.
+func (auth *AuthManager) handleReinstate(conn comms.Link, msg *msgjson.Message) *msgjson.Error {
+	// Unmarshal.
+	reinstate := new(msgjson.Reinstate)
+	err := json.Unmarshal(msg.Payload, &reinstate)
+	if err != nil {
+		return &msgjson.Error{
+			Code:    msgjson.RPCParseError,
+			Message: "error parsing reinstate: " + err.Error(),
+		}
+	}
+
+	pubKey, err := secp256k1.ParsePubKey(reinstate.ClientPubKey)
+	if err != nil {
+		return &msgjson.Error{
+			Code:    msgjson.PubKeyParseError,
+			Message: "error parsing pubkey: " + err.Error(),
+		}
+	}
+
+	acct := &account.Account{
+		ID:     account.NewID(reinstate.ClientPubKey),
+		PubKey: pubKey,
+	}
+
+	// Check signature.
+	sigMsg := reinstate.Serialize()
+	err = checkSigS256(sigMsg, reinstate.SigBytes(), acct.PubKey)
+	if err != nil {
+		return &msgjson.Error{
+			Code:    msgjson.SignatureError,
+			Message: "signature error: " + err.Error(),
+		}
+	}
+
+	// Decode account proof.
+	accountProof, err := msgjson.DecodeAccountProof(reinstate.AccountProof)
+	if err != nil {
+		return &msgjson.Error{
+			Code:    msgjson.InvalidAccountProof,
+			Message: "DecodeAccountProof error: " + err.Error(),
+		}
+	}
+
+	// Verify NotifyFee receipt
+	notifyFee := &msgjson.NotifyFee{
+		AccountID: acct.ID[:],
+		CoinID:    reinstate.CoinID,
+		Time:      accountProof.Stamp,
+	}
+	notifyFee.Sig = accountProof.Sig
+	err = checkSigS256(notifyFee.Serialize(), accountProof.Sig, auth.signer.PubKey())
```

Comment:
Oh, I see... the idea is to have the server's original signature of the notify fee message prove that the fee went to the address provided to the client initially.  Interesting, will dig in and get a review.